### PR TITLE
feat: improve Github actions

### PR DESCRIPTION
**Pull request by JoeZwet**

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -1,24 +1,43 @@
...
   build:
-    runs-on: ubuntu-20.04
-    container:
-      image: openjdk:8-jdk
+    strategy:
+      matrix:
+        java: [
+          1.8, # Minimum
+          11,  # LTS
+          16,  # Latest
```
Comment (marcus8448) [inline .github/workflows/build.yml:14]: Do we really need to test on all these versions? Minimum should be good enough, or even minimum and latest, if there's no forward compatibility for some reason.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -1,24 +1,43 @@
...
+        run: chmod +x ./gradlew
+      - name: Build
         uses: eskatos/gradle-command-action@v1
         with:
           arguments: checkLicenses build
           wrapper-cache-enabled: true
           dependencies-cache-enabled: true
-      - name: capture build artifacts
+      - name: Capture build artifacts
+        if: ${{ runner.os == 'Linux' && matrix.java == '11' }}
```
Comment (marcus8448) [inline .github/workflows/build.yml:39]: Maybe capture artifacts for 8 - there *technically* is a backwards compatibility issue with `FloatBuffer#flip` (signature changed [but also somehow forwards compatible]), not that we're using that yet.

File: .github/workflows/pr_check.yml [hunk 3]
```diff
@@ -1,19 +1,34 @@
...
-      - name: validate gradle wrapper
+        with:
+          fetch-depth: 0
+      - name: Validate Gradle wrapper
         uses: gradle/wrapper-validation-action@v1
-      - name: build
+      - name: Setup JDK ${{ matrix.java }}
+        uses: actions/setup-java@v1
+        with:
+          java-version: ${{ matrix.java }}
```
Comment (marcus8448) [inline .github/workflows/pr_check.yml:27]:
```suggestion
        uses: actions/setup-java@v2
        with:
          java-version: ${{ matrix.java }}
          distribution: 'adopt'
```

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -1,24 +1,43 @@
...
-      - name: validate gradle wrapper
+        with:
+          fetch-depth: 0
+      - name: Validate Gradle wrapper
         uses: gradle/wrapper-validation-action@v1
-      - name: build
+      - name: Setup JDK ${{ matrix.java }}
+        uses: actions/setup-java@v1
+        with:
+          java-version: ${{ matrix.java }}
```
Comment (marcus8448) [inline .github/workflows/build.yml:28]:
```suggestion
        uses: actions/setup-java@v2
        with:
          java-version: ${{ matrix.java }}
          distribution: 'adopt'
```

File: .github/workflows/pr_check.yml [hunk 5]
```diff
@@ -1,19 +1,34 @@
...
 
 jobs:
   build:
-    runs-on: ubuntu-20.04
-    container:
-      image: openjdk:8-jdk
+    strategy:
+      matrix:
+        java: [
+            1.8, # Minimum
```
Comment (marcus8448) [inline .github/workflows/pr_check.yml:11]:
```suggestion
            8, # Minimum
```
Required for `actions/setup-java@v2`

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -1,24 +1,43 @@
...
 
 jobs:
   build:
-    runs-on: ubuntu-20.04
-    container:
-      image: openjdk:8-jdk
+    strategy:
+      matrix:
+        java: [
+          1.8, # Minimum
```
Comment (marcus8448) [inline .github/workflows/build.yml:12]:
```suggestion
          8, # Minimum
```
Required for `actions/setup-java@v2`

File: .github/workflows/publish.yml [hunk 7]
```diff
@@ -2,20 +2,28 @@ name: Maven Publish
...
+      - name: Checkout repository
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+      - name: Validate Gradle wrapper
+        uses: gradle/wrapper-validation-action@v1
+      - name: Setup JDK 1.8
+        uses: actions/setup-java@v1
         with:
           java-version: 1.8
```
Comment (marcus8448) [inline .github/workflows/publish.yml:20]:
```suggestion
        uses: actions/setup-java@v2
        with:
          java-version: 8
          distribution: 'adopt'
```

**Review by marcus8448 — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml (hunk 1 above)
Comment (JoeZwet) [reply .github/workflows/build.yml]: iirc most people will use lts for custom installs ie multimc

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -1,24 +1,44 @@
...
-    container:
-      image: openjdk:8-jdk
+    strategy:
+      matrix:
+        java: [
+          8, # Minimum
+          11,  # LTS
+          16,  # Latest
+        ]
+        os: [ubuntu-20.04]
```
Comment (marcus8448) [inline .github/workflows/build.yml:16]: There's no reason to have this if the array only contains one entry. (Isn't java meant to be 'write once, run anywhere' anyways?)

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -1,24 +1,44 @@
...
+      - name: Validate Gradle wrapper
         uses: gradle/wrapper-validation-action@v1
-      - name: build
+      - name: Setup JDK ${{ matrix.java }}
+        uses: actions/setup-java@v2
+        with:
+          java-version: ${{ matrix.java }}
+          distribution: 'adopt'
+      - name: Make Gradle wrapper executable
+        if: ${{ runner.os != 'Windows' }}
```
Comment (marcus8448) [inline .github/workflows/build.yml:31]: We never actually run on windows

File: .github/workflows/build.yml (hunk 2 above)
Comment (JoeZwet) [reply .github/workflows/build.yml]: resolved in c02a824

File: .github/workflows/build.yml [hunk 10]
```diff
@@ -1,24 +1,44 @@
...
+        run: chmod +x ./gradlew
+      - name: Build
         uses: eskatos/gradle-command-action@v1
         with:
           arguments: checkLicenses build
           wrapper-cache-enabled: true
           dependencies-cache-enabled: true
-      - name: capture build artifacts
+      - name: Capture build artifacts
+        if: ${{ runner.os == 'Linux' && matrix.java == '11' }}
```
Comment (marcus8448) [inline .github/workflows/build.yml]: We always run on linux

File: .github/workflows/pr_check.yml [hunk 11]
```diff
@@ -1,21 +1,37 @@
...
-    container:
-      image: openjdk:8-jdk
+    strategy:
+      matrix:
+        java: [
+            8, # Minimum
+            11,  # LTS
+            16,  # Latest
+        ]
+        os: [ubuntu-20.04]
```
Comment (marcus8448) [inline .github/workflows/pr_check.yml:15]: Same thing - There's no reason to have this if we're only using ubuntu.

File: .github/workflows/pr_check.yml [hunk 12]
```diff
@@ -1,21 +1,37 @@
...
+      - name: Validate Gradle wrapper
         uses: gradle/wrapper-validation-action@v1
-      - name: build
+      - name: Setup JDK ${{ matrix.java }}
+        uses: actions/setup-java@v2
+        with:
+          java-version: ${{ matrix.java }}
+          distribution: 'adopt'
+      - name: Make Gradle wrapper executable
+        if: ${{ runner.os != 'Windows' }}
```
Comment (marcus8448) [inline .github/workflows/pr_check.yml:30]: We don't run actions on Windows.

File: .github/workflows/publish.yml [hunk 13]
```diff
@@ -2,20 +2,29 @@ name: Maven Publish
...
+          fetch-depth: 0
+      - name: Validate Gradle wrapper
+        uses: gradle/wrapper-validation-action@v1
+      - name: Setup JDK 1.8
+        uses: actions/setup-java@v2
+        with:
+          java-version: 8
+          distribution: 'adopt'
+      - name: Make Gradle wrapper executable
+        if: ${{ runner.os != 'Windows' }}
```
Comment (marcus8448) [inline .github/workflows/publish.yml:23]: We don't run this on Windows (this action doesn't have a os matrix either)

**Review by marcus8448 — CHANGES_REQUESTED** [review]

File: .github/workflows/publish.yml (hunk 13 above)
Comment (JoeZwet) [reply .github/workflows/publish.yml:23]: its more of a sanity check for if we ever do

File: .github/workflows/build.yml (hunk 1 above)
Comment (marcus8448) [reply .github/workflows/build.yml]: What I meant is: Is it necessary to test compilation on all three versions? Java *shouldn't* be breaking old things with new releases (Unless we're using unsupported APIs like `Unsafe`).

File: .github/workflows/build.yml (hunk 8 above)
Comment (JoeZwet) [reply .github/workflows/build.yml:16]: yeah, but this is just future proofing it, in case we ever do want to build on windows its already set up to do so.

File: .github/workflows/build.yml (hunk 9 above)
Comment (JoeZwet) [reply .github/workflows/build.yml:31]: sanity check for if we ever do

File: .github/workflows/build.yml [hunk 14]
```diff
@@ -1,24 +1,44 @@
...
         with:
           arguments: checkLicenses build
           wrapper-cache-enabled: true
           dependencies-cache-enabled: true
-      - name: capture build artifacts
+      - name: Capture build artifacts
+        if: ${{ runner.os == 'Linux' && matrix.java == '8' }}
         uses: actions/upload-artifact@v2
         with:
           name: Artifacts
```
Comment (marcus8448) [inline .github/workflows/build.yml:43]:
```suggestion
          name: Galacticraft
```

File: .github/workflows/build.yml [hunk 15]
```diff
@@ -1,24 +1,44 @@
...
-          path: build/libs/
+          path: build/libs
```
Comment (marcus8448) [inline .github/workflows/build.yml:44]:
```suggestion
          path: |
            build/libs/
            !build/libs/*-sources.jar
            !build/libs/*-dev.jar
```

**Review by marcus8448 — CHANGES_REQUESTED** [review]
Might as well also toss this in

File: .github/workflows/build.yml (hunk 1 above)
Comment (JoeZwet) [reply .github/workflows/build.yml]: it's probably not needed but it's a good sanity check so we know it works on multiple versions.

**Review by marcus8448 — APPROVED** [review]
